api/upload.py

```diff
@@ -1,18 +1,130 @@
 import bson
-import os.path
 import datetime
+import json
+import os.path
 
 from . import base
-from . import util
+from . import config
 from . import files
 from . import rules
-from . import config
-from .dao import reaperutil, APIStorageException
-from . import validators
 from . import tempdir as tempfile
+from . import placer as pl
+from . import util
+from . import validators
+from .dao import reaperutil, APIStorageException
 
 log = config.log
 
+Strategy = util.Enum('Strategy', {
+    'targeted' : pl.TargetedPlacer,   # Upload N files to a container.
+    'connector': pl.ConnectorPlacer,  # Upload N files from a scientific data source.
+    'engine'   : pl.EnginePlacer,	 # Upload N files from the result of a successful job.
+    'packfile' : pl.PackfilePlacer	# Upload N files as a new packfile to a container.
+})
+
+def process_upload(request, strategy, container_type=None, id=None):
+    """
+    Universal file upload entrypoint.
+
+    Format:
+        Multipart form upload with N file fields, each with their desired filename.
+        For technical reasons, no form field names can be repeated. Instead, use (file1, file2) and so forth.
+
+        Depending on the type of upload, a non-file form field called "metadata" may/must also be sent.
+        If present, it is expected to be a JSON string matching the schema for the upload strategy.
+
+        Currently, the JSON returned may vary by strategy.
+
+        Some examples:
+        curl -F file1=@science.png   -F file2=@rules.png url
+        curl -F metadata=<stuff.json -F file=@data.zip   url
+        http --form POST url metadata=@stuff.json file@data.zip
+
+    Features:
+                                               | targeted | connector | engine | packfile
+        Must specify a target container        |     X    |           |    X   |
+        May create hierarchy on demand         |          |     X     |        |     X
+
+        May  send metadata about the files     |     X    |     X     |    X   |     X
+        MUST send metadata about the files     |          |     X     |        |     X
+
+        Creates a packfile from uploaded files |          |           |        |     X
+    """
```

Comment:
"connector" is a Flywheel marketing term. The correct scitran terminology is "reaper". Please change globally.